### DISCUSSION TITLE
Pointer type usage like: ①pointer to pointer ② reference of pointer

### DISCUSSION THREAD

**Discussion by lyd405121** [wjakob/nanobind#447] [Q&A] [answered]
## Background
- Recently，I build a reflect system for C++
- Using Clang to parse *.h files and automaticlly output [rttr](https://github.com/rttrorg/rttr) code 
- And I suddenly came up an idea: **Can I using that system to produce python glue code with nanobind**
- Input any *.h and output nanobind extension automaticly was my ambition！
## Question
- It is very easy to achieve that when the input includes only data oriented struct
- When it turns to interface layer, it's really hard to deal with **double pointer** and **reference of pointer**
![PoiterTestCode](https://github.com/wjakob/nanobind/assets/16344694/b569c10e-6d7a-490f-b8f6-769d83067f00)
- nanobind can't work with double pointer(**Test4**) and reference of pointer(**Test3**), so I use lambda function instead
- Which successfully compiled and my python code below also print the address of the pointer  
![PoiterTestPy](https://github.com/wjakob/nanobind/assets/16344694/a806b459-357e-43d9-b9d2-7506bb377fb0)
- But which confuse me is some addtional print info :
![PoiterTestRet](https://github.com/wjakob/nanobind/assets/16344694/dfb30528-3729-4951-8e22-9b0bf3d43733)

### Question 1 : Two address
- Why there are two address for the variant “value=-1”: 7db25ef1d4 and 7db25ef1f4
- Why the distance between the two address is 32 

### Question 2 : Python value can't be modified
- The variant "value = -1" never changed  even I use its pointer to modify

## Guess
- nanobind store parameter list of a function globally, and my code modify the data in nanobind not the python variant
- I hope my Chinglish has not affected your reading T_T, I just want to figure out nanobind's strategy for pointer

**Comment by wjakob** [comment, marked as answer]
Please don't file issues for things that aren't bugs in the library. I will move this to the discussion page. I cannot provide support for what you are trying to do, but perhaps others have ideas.

**Comment by oremanj** [comment]
There is no reasonable way for nanobind to support double indirection here, because the "outer" pointer or reference needs to be able to point at an "inner" pointer at some fixed location in memory, and that doesn't exist.

* For a type caster, which is e.g. loading an `int` from Python, nanobind converts the Python `int` to a C++ integer type which is stored on the stack (inside the type_caster object) during the function call that needs it. You can make one level of pointer to that. But that first level of pointer isn't stored anywhere, so you can't in turn point to it.
* Even for a bound type, the C++ object is stored inline in the Python object, so it has a well-defined address but that address isn't stored anywhere.

You could jump through hoops with a custom type_caster or something to support this, but it would be kind of useless, because you can't actually *use* the double-pointer to change what the inside pointer points to, which is the main reason someone would invoke one of these exotic creatures in the first place.

**Reply by lyd405121** [reply]
## Thanks
- Thanks for your reply, completely get that
- Now I am using pointer-int cast to register object address outside nanobind

## My solution
- Cpp code to register long long pointer as input for python script

```cpp
    //nanobind code generated by custom clang tool
    NB_MODULE(HKSceneGraphPy, m)
    {
        m.def("SetResourceDir", &SetResourceDir);
        m.def("SetCurrentScene", &SetCurrentScene);


        nb::class_<HKSceneGraph>(m,"HKSceneGraph")
        .def(nb::init<>())
        .def(nb::init<bool>())
        .def("Version", &HKSceneGraph::Version)
        .def("Load", &HKSceneGraph::Load)
        .def("Save", &HKSceneGraph::Save)
        .def("Add", nb::overload_cast<SG_ASSET&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_CAMERA&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_LIGHT&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_MATERIAL&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_MORPH_ANIM&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_SKIN_ANIM&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_INSTANCE&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_LIGHT&>(&HKSceneGraph::Add))
        .def("Add", nb::overload_cast<SG_TEXTURE&>(&HKSceneGraph::Add))
    }

    // client cpp
    PyObject* pLoadModule = PyImport_ImportModule("RuntimeLoadScene");
    PyObject* LoadFunc = PyObject_GetAttrString(pLoadModule, "Script"); 
    PyObject* pArgs    = PyTuple_New(1);
    PyTuple_SetItem(pArgs, 0, Py_BuildValue("L", (long long)pSceneGraph));
    PyObject_CallObject(LoadFunc, pArgs);    
```

- Python code recieve the long long value as current_scene 
- Cast the long long value to the address using nanobind generated module

```python
import sys
import os
import HKSceneGraphPy as SG

def Script(current_scene):
    SG.SetResourceDir(os.environ.get('HK_SAMPLE_RESOURCE_DIR'))
    SG.SetCurrentScene(current_scene)
    scene = SG.HKSceneGraph(True)
    scene.Load("Barcode.scene")
```
